pkg/parser/types/field_type.go

```diff
@@ -61,6 +76,68 @@ type FieldType struct {
 	// Please keep in mind that jsonFieldType should be updated if you add a new field here.
 }
 
+// Hash64 implements the cascades/base.Hasher.<0th> interface.
+func (ft *FieldType) Hash64(h IHasher) {
+	h.HashByte(ft.tp)
+	h.HashUint64(uint64(ft.flag))
+	h.HashInt(ft.flen)
+	h.HashInt(ft.decimal)
+	h.HashString(ft.charset)
+	h.HashString(ft.collate)
+	h.HashInt(len(ft.elems))
+	for _, elem := range ft.elems {
+		h.HashString(elem)
+	}
+	h.HashInt(len(ft.elemsIsBinaryLit))
+	for _, elem := range ft.elemsIsBinaryLit {
+		h.HashBool(elem)
+	}
+	h.HashBool(ft.array)
+}
+
+// Equals implements the cascades/base.Hasher.<1th> interface.
+func (ft *FieldType) Equals(other any) bool {
+	if other == nil {
+		return false
+	}
+	var ft2 *FieldType
+	switch x := other.(type) {
+	case *FieldType:
+		ft2 = x
+	case FieldType:
+		ft2 = &x
+	default:
+		return false
+	}
+	ok := ft.tp == ft2.tp &&
+		ft.flag == ft2.flag &&
+		ft.flen == ft2.flen &&
+		ft.decimal == ft2.decimal &&
+		ft.charset == ft2.charset &&
+		ft.collate == ft2.collate &&
+		ft.array == ft2.array
+	if !ok {
+		return false
+	}
+	if len(ft.elems) != len(ft2.elems) {
+		return false
+	}
+	for i, one := range ft.elems {
+		if one != ft2.elems[i] {
```

Comment:
do we need to consider collation here?

Reply:
but if you update the `hash64` method to consider the collation, it is possible that under ci collation, two type with "a" and "A" is equal and has the same hash6. So the problem is, by you original define and purpose, enum should we define type("a","b") euqal to enum type("A","B")  when collation is ci?

Reply:
“by you original definition and purpose, should we define enum type("a","b") equals to enum type("A","B") when collation is ci?” 
this is not in consideration here, high-level equivalence is guaranteed by datum-compare liked infra, hash64 is based from the basic representation of data structure.